theano/sandbox/cuda/acdc.py

```diff
@@ -0,0 +1,309 @@
+#
+# The following is based on acdc_wrapper.cu
+# from https://github.com/mdenil/acdc-torch
+#
+
+import os
+
+import theano
+from theano.sandbox.cuda import GpuOp
+from theano.sandbox.cuda.basic_ops import as_cuda_ndarray_variable, gpu_contiguous
+
+import cuda_ndarray
+
+class FastACDC(GpuOp):
+
+    def c_headers(self):
+        return ['<stdio.h>', '<cufft.h>']
+
+    def c_libraries(self):
+        return ['cufft']
+
+    def make_node(self, x, A, Ab, D, Db):
+        x_ = gpu_contiguous(as_cuda_ndarray_variable(x))
+        A_ = gpu_contiguous(as_cuda_ndarray_variable(A))
+        Ab_ = gpu_contiguous(as_cuda_ndarray_variable(Ab))
+        D_ = gpu_contiguous(as_cuda_ndarray_variable(D))
+        Db_ = gpu_contiguous(as_cuda_ndarray_variable(Db))
+        for t in [x_, A_, Ab_, D_, Db_]:
+            assert t.dtype == 'float32'
+        return theano.Apply(self, [x_, A_, Ab_, D_, Db_], [x_.type()])
+
+    def grad(self, inputs, cost_grad):
+        inp, A, Ab, D, Db = inputs
+        gradOutput, = cost_grad
+        gradOutput = gpu_contiguous(gradOutput)
+        return FastACDCGrad()(inp, gradOutput, A, Ab, D, Db)
```

Comment:
building  a python object is a little costly, so we use the pattern at many place. This also show the user a better interface that don't request them to instantiate an op.

class MyOp(Op):
    ...
my_op = MyOp()

then people can use my_op(some_theano_variables, ...)

You can do this for both op and reuse the fast_acdc_grad instance here.